addons/hr_payroll/models/hr_leave.py

```diff
@@ -13,6 +16,36 @@ class HrLeaveType(models.Model):
 class HrLeave(models.Model):
     _inherit = 'hr.leave'
 
+    @api.multi
+    def _create_resource_leave(self):
+        """
+            We override this method to add the leave in the calendars of contracts.
+            Its needed when generating payslips to have leave hours corresponding
+            to the working hours of contracts.
+        """
+        super(HrLeave, self)._create_resource_leave()
+        values = []
```

Comment:
Careful, a leave might not have an employee linked to it

ps : I find something like this clearer
```
values = []

for leave in self:
    if leave.employee_id:
        vals = {
          'name': leave.name,
          'date_from': leave.date_from,
          'holiday_id': leave.id,
          'date_to': leave.date_to,
          'resource_id': leave.employee_id.resource_id.id,
          'time_type': leave.holiday_status_id.time_type
        }

        calendars = leave.employee_id._get_contracts(leave.date_from, leave.date_to).mapped('resource_calendar_id')
        values += [dict(vals, calendar_id=calendar.id) for calendar in calendars]

self.env['resource.calendar.leaves'].create(values)
```